protocol/v1/qbft/validation/changeround/validate_justification.go

```diff
@@ -38,67 +43,85 @@ func (p *validateJustification) Run(signedMessage *specqbft.SignedMessage) error
+	if err := data.Validate(); err != nil {
+		return fmt.Errorf("roundChangeData invalid: %w", err)
 	}
 
-	// validateJustification justification signature
-	pksMap, err := p.share.PubKeysByID(data.RoundChangeJustification[0].GetSigners())
-	var pks beacon.PubKeys
-	for _, v := range pksMap {
-		pks = append(pks, v)
+	if !data.Prepared() {
+		return nil
 	}
 
-	if err != nil {
-		return errors.Wrap(err, "change round could not get pubkey")
-	}
-	aggregated := pks.Aggregate()
-	for _, justification := range data.RoundChangeJustification {
-		err = justification.Signature.Verify(justification, types.GetDefaultDomain(), spectypes.QBFTSignatureType, aggregated.Serialize())
+	for _, rcj := range data.RoundChangeJustification {
+		if rcj.Message == nil {
+			return errors.New("change round justification msg is nil")
+		}
+		if rcj.Message.MsgType != specqbft.PrepareMsgType {
+			return errors.Errorf("change round justification msg type not Prepare (%d)", rcj.Message.MsgType)
+		}
+		if signedMessage.Message.Height != rcj.Message.Height {
+			return errors.New("change round justification sequence is wrong")
+		}
+		if signedMessage.Message.Round <= rcj.Message.Round {
+			return errors.New("round change justification invalid: msg round wrong")
+		}
+		if data.PreparedRound != rcj.Message.Round {
+			return errors.New("round change justification invalid: msg round wrong")
+		}
+		if !bytes.Equal(signedMessage.Message.Identifier, rcj.Message.Identifier) {
+			return errors.New("change round justification msg Lambda not equal to msg Lambda not equal to instance lambda")
+		}
+		prepareMsg, err := rcj.Message.GetPrepareData()
+		if err != nil {
+			return errors.Wrap(err, "failed to get prepare data")
+		}
+		if !bytes.Equal(data.PreparedValue, prepareMsg.Data) {
+			return errors.New("round change justification invalid: prepare data != proposed data")
+		}
+		if len(rcj.GetSigners()) != 1 {
+			return errors.New("round change msg allows 1 signer")
+		}
+
+		// validateJustification justification signature
+		pksMap, err := p.share.PubKeysByID(rcj.GetSigners())
+		var pks beacon.PubKeys
+		for _, v := range pksMap {
+			pks = append(pks, v)
+		}
+
 		if err != nil {
+			return errors.Wrap(err, "change round could not get pubkey")
+		}
+		aggregated := pks.Aggregate()
+
+		if err = rcj.Signature.Verify(rcj, types.GetDefaultDomain(), spectypes.QBFTSignatureType, aggregated.Serialize()); err != nil {
 			return errors.Wrap(err, "change round could not verify signature")
 		}
 	}
 
+	if quorum, _, _ := p.changeRoundQuorum(data.RoundChangeJustification); !quorum {
+		return fmt.Errorf("no justifications quorum")
+	}
+
+	if data.PreparedRound > signedMessage.Message.Round {
+		return errors.New("prepared round > round")
+	}
+
 	return nil
 }
 
 // Name implements pipeline.Pipeline interface
 func (p *validateJustification) Name() string {
 	return "validateJustification msg"
 }
+
+// TODO(nkryuchkov): Consider merging with all changeRoundQuorum functions.
```

Comment:
lets merge it. 
this function can receive share var instead of `p *validateJustification`.
keep it in validateJustification file

Reply:
@nivBlox Done